src/core/vectortile/qgsvectortilemvtencoder.cpp

```diff
@@ -0,0 +1,357 @@
+    if ( !layerTileExtent.intersects( layer->extent() ) )
+    {
+      return;  // tile is completely outside of the layer'e extent
+    }
+  }
+  catch ( const QgsCsException & )
+  {
+    QgsDebugMsg( "Failed to reproject tile extent to the layer" );
+    return;
+  }
+
+  if ( layerName.isEmpty() )
+    layerName = layer->name();
+
+  // add buffer to both filter extent in layer CRS (for feature request) and tile extent in target CRS (for clipping)
+  double bufferRatio = static_cast<double>( mBuffer ) / mResolution;
+  QgsRectangle tileExtent = mTileExtent;
+  tileExtent.grow( bufferRatio * mTileExtent.width() );
+  layerTileExtent.grow( bufferRatio * std::max( layerTileExtent.width(), layerTileExtent.height() ) );
+
+  QgsFeatureRequest request;
+  request.setFilterRect( layerTileExtent );
+  if ( !filterExpression.isEmpty() )
+    request.setFilterExpression( filterExpression );
+  QgsFeatureIterator fit = layer->getFeatures( request );
+
+  QgsFeature f;
+  if ( !fit.nextFeature( f ) )
+  {
+    return;  // nothing to write - do not add the layer at all
+  }
+
+  vector_tile::Tile_Layer *tileLayer = tile.add_layers();
+  tileLayer->set_name( layerName.toUtf8() );
+  tileLayer->set_version( 2 );  // 2 means MVT spec version 2.1
+  tileLayer->set_extent( static_cast<::google::protobuf::uint32>( mResolution ) );
+
+  const QgsFields fields = layer->fields();
+  for ( int i = 0; i < fields.count(); ++i )
+  {
+    tileLayer->add_keys( fields[i].name().toUtf8() );
+  }
+
+  do
+  {
+    QgsGeometry g = f.geometry();
+
+    // reproject
+    try
+    {
+      g.transform( ct );
+    }
+    catch ( const QgsCsException & )
+    {
+      QgsDebugMsg( "Failed to reproject geometry " + QString::number( f.id() ) );
+      continue;
+    }
+
+    // clip
+    g = g.clipped( tileExtent );
+
+    f.setGeometry( g );
+
+    addFeature( tileLayer, f );
+  }
+  while ( fit.nextFeature( f ) );
+
+  mKnownValues.clear();
+}
+
+void QgsVectorTileMVTEncoder::addFeature( vector_tile::Tile_Layer *tileLayer, const QgsFeature &f )
+{
+  vector_tile::Tile_Feature *feature = tileLayer->add_features();
+
+  feature->set_id( static_cast<quint64>( f.id() ) );
+
+  //
+  // encode attributes
+  //
+
+  const QgsAttributes attrs = f.attributes();
+  for ( int i = 0; i < attrs.count(); ++i )
+  {
+    const QVariant v = attrs.at( i );
+    if ( !v.isValid() || v.isNull() )
+      continue;
+
+    int valueIndex;
+    if ( mKnownValues.contains( v ) )
+    {
+      valueIndex = mKnownValues[v];
+    }
+    else
+    {
+      vector_tile::Tile_Value *value = tileLayer->add_values();
+      valueIndex = tileLayer->values_size() - 1;
+      mKnownValues[v] = valueIndex;
+
+      if ( v.type() == QVariant::Double )
+        value->set_double_value( v.toDouble() );
+      else if ( v.type() == QVariant::Int )
+        value->set_int_value( v.toInt() );
+      else if ( v.type() == QVariant::Bool )
+        value->set_bool_value( v.toBool() );
+      else
```

Comment:
Are other types allowed in the spec? dates/times?

Reply:
Not really, just strings, floats/doubles, ints, bools

https://github.com/mapbox/vector-tile-spec/blob/master/2.1/vector_tile.proto#L17